File: src-windows/CRT.Core/Net/InnertubeClient.cs
using System.Collections.Concurrent;
using System.Globalization;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace CRT.Core.Net;

/// <summary>
/// Looks up the true encoded framerate of a YouTube video/itag pair via the
/// Innertube player API (no yt-dlp dependency), with an optional yt-dlp
/// fallback. Returns null on any failure — callers treat that as "couldn't be
/// verified" and never block on it. Results are cached per (videoId, itag) for
/// the app session.
/// </summary>
public sealed class InnertubeClient
{
    private readonly HttpClient _http;
    private readonly Func<string, string, CancellationToken, Task<decimal?>>? _ytDlpFallback;
    private readonly ConcurrentDictionary<(string VideoId, string FormatId), decimal?> _cache = new();

    /// <param name="ytDlpFallback">
    /// Optional fallback invoked as (videoId, formatId) → fps when the
    /// Innertube call fails and a yt-dlp binary is available.
    /// </param>
    public InnertubeClient(HttpClient? http = null, Func<string, string, CancellationToken, Task<decimal?>>? ytDlpFallback = null)
    {
        _http = http ?? new HttpClient();
        _http.Timeout = TimeSpan.FromSeconds(8);
        _ytDlpFallback = ytDlpFallback;
    }

    public async Task<decimal?> GetFormatFramerateAsync(string videoId, string formatId, CancellationToken ct = default)
    {
        var key = (videoId, formatId);
        if (_cache.TryGetValue(key, out decimal? cached))
        {
            return cached;
        }

        decimal? result = await LookupAsync(videoId, formatId, ct).ConfigureAwait(false);
        _cache[key] = result;
        return result;
    }

    private async Task<decimal?> LookupAsync(string videoId, string formatId, CancellationToken ct)
    {
        decimal? fps =
            await TryPlayerRequestAsync(videoId, formatId, BuildAndroidBody(videoId), InnertubeClients.AndroidUserAgent, ct).ConfigureAwait(false)
            ?? await TryPlayerRequestAsync(videoId, formatId, BuildIosBody(videoId), InnertubeClients.IosUserAgent, ct).ConfigureAwait(false);
        if (fps is not null)
        {
            return fps;
        }

        if (_ytDlpFallback is not null)
        {
            try
            {
                return await _ytDlpFallback(videoId, formatId, ct).ConfigureAwait(false);
            }
            catch (Exception e) when (e is not OperationCanceledException)
            {
                return null;
            }
        }
        return null;
    }

    private static string BuildAndroidBody(string videoId) => JsonSerializer.Serialize(new
    {
        videoId,
        context = new
        {
            client = new
            {
                clientName = InnertubeClients.AndroidClientName,
                clientVersion = InnertubeClients.AndroidClientVersion,
                androidSdkVersion = InnertubeClients.AndroidSdkVersion,
                hl = "en",
            },
        },
    });

    private static string BuildIosBody(string videoId) => JsonSerializer.Serialize(new
    {
        videoId,
        context = new
        {
            client = new
            {
                clientName = InnertubeClients.IosClientName,
                clientVersion = InnertubeClients.IosClientVersion,
                deviceModel = InnertubeClients.IosDeviceModel,
                hl = "en",
            },
        },
    });

    private async Task<decimal?> TryPlayerRequestAsync(string videoId, string formatId, string body, string userAgent, CancellationToken ct)
    {
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, InnertubeClients.PlayerEndpoint)
            {
                Content = new StringContent(body, Encoding.UTF8, "application/json"),
            };
            request.Headers.TryAddWithoutValidation("User-Agent", userAgent);

            using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                return null;
            }

            using var document = JsonDocument.Parse(await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false));
            if (!document.RootElement.TryGetProperty("streamingData", out JsonElement streamingData))
            {
                return null;
            }

            return FindItagFps(streamingData, "formats", formatId)
                ?? FindItagFps(streamingData, "adaptiveFormats", formatId);
        }
        catch (Exception e) when (e is HttpRequestException or TaskCanceledException or JsonException)
        {
            return null;
        }
    }

    private static decimal? FindItagFps(JsonElement streamingData, string listName, string formatId)
    {
        if (!streamingData.TryGetProperty(listName, out JsonElement formats) || formats.ValueKind != JsonValueKind.Array)
        {
            return null;
        }

        foreach (var format in formats.EnumerateArray())
        {
            if (!format.TryGetProperty("itag", out JsonElement itag))
            {
                continue;
            }
            // "fmt" from debug info is a string — compare after string conversion.
            string itagText = itag.ValueKind == JsonValueKind.Number ? itag.GetRawText() : itag.GetString() ?? "";
            if (itagText != formatId)
            {
                continue;
            }
            if (format.TryGetProperty("fps", out JsonElement fps) && fps.ValueKind == JsonValueKind.Number)
            {
                return decimal.Parse(fps.GetRawText(), CultureInfo.InvariantCulture);
            }
            return null;
        }
        return null;
    }
}

File: src-macos/Sources/CRTCore/InnertubeClient.swift
import Foundation

/// YouTube innertube client identities. YouTube rotates accepted client
/// versions periodically — update these constants (only) when lookups start
/// returning HTTP 400 FAILED_PRECONDITION.
public enum InnertubeClientIdentity {
    /// Primary: ANDROID client (verified working 2026-08).
    public static let androidClientName = "ANDROID"
    public static let androidClientVersion = "20.10.38"
    public static let androidSdkVersion = 30
    public static let androidUserAgent = "com.google.android.youtube/20.10.38 (Linux; U; Android 11) gzip"

    /// Fallback: IOS client. Never use the WEB client (zero formats).
    public static let iosClientName = "IOS"
    public static let iosClientVersion = "20.10.4"
    public static let iosDeviceModel = "iPhone16,2"
    public static let iosUserAgent = "com.google.ios.youtube/20.10.4 (iPhone16,2; U; CPU iOS 17_5_1 like Mac OS X)"
}

/// Looks up the real encoded framerate of a YouTube video format via the
/// innertube player endpoint (spec §2.1). All failures return nil — callers
/// treat that as "couldn't be verified". Results (including failures) are
/// cached per (videoId, itag) for the app session.
public actor InnertubeClient {
    public static let shared = InnertubeClient()

    private var cache: [String: Decimal?] = [:]
    private let session: URLSession

    public init() {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 8
        configuration.timeoutIntervalForResource = 8
        self.session = URLSession(configuration: configuration)
    }

    /// The cached lookup. `formatID` is the debug info's `fmt` value (string).
    public func framerate(videoID: String, formatID: String) async -> Decimal? {
        let key = "\(videoID)|\(formatID)"
        if let cached = cache[key] {
            return cached
        }
        let result = await lookup(videoID: videoID, formatID: formatID)
        cache[key] = result
        return result
    }

    private func lookup(videoID: String, formatID: String) async -> Decimal? {
        if let fps = await requestPlayer(videoID: videoID, formatID: formatID, useIOSClient: false) {
            return fps
        }
        return await requestPlayer(videoID: videoID, formatID: formatID, useIOSClient: true)
    }

    private func requestPlayer(videoID: String, formatID: String, useIOSClient: Bool) async -> Decimal? {
        guard let url = URL(string: "https://www.youtube.com/youtubei/v1/player") else {
            return nil
        }

        var client: [String: Any]
        let userAgent: String
        if useIOSClient {
            client = [
                "clientName": InnertubeClientIdentity.iosClientName,
                "clientVersion": InnertubeClientIdentity.iosClientVersion,
                "deviceModel": InnertubeClientIdentity.iosDeviceModel,
                "hl": "en",
            ]
            userAgent = InnertubeClientIdentity.iosUserAgent
        } else {
            client = [
                "clientName": InnertubeClientIdentity.androidClientName,
                "clientVersion": InnertubeClientIdentity.androidClientVersion,
                "androidSdkVersion": InnertubeClientIdentity.androidSdkVersion,
                "hl": "en",
            ]
            userAgent = InnertubeClientIdentity.androidUserAgent
        }

        let body: [String: Any] = [
            "videoId": videoID,
            "context": ["client": client],
        ]

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        guard let bodyData = try? JSONSerialization.data(withJSONObject: body) else {
            return nil
        }
        request.httpBody = bodyData

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                return nil
            }
            return InnertubeClient.parsePlayerResponse(data, formatID: formatID)
        } catch {
            return nil
        }
    }

    /// Reads `streamingData.formats` + `streamingData.adaptiveFormats`, finds
    /// the entry whose itag matches (string-compared), returns its fps.
    public static func parsePlayerResponse(_ data: Data, formatID: String) -> Decimal? {
        guard
            let object = try? JSONSerialization.jsonObject(with: data),
            let dict = object as? [String: Any],
            let streamingData = dict["streamingData"] as? [String: Any]
        else {
            return nil
        }

        var formats: [[String: Any]] = []
        if let plain = streamingData["formats"] as? [[String: Any]] {
            formats.append(contentsOf: plain)
        }
        if let adaptive = streamingData["adaptiveFormats"] as? [[String: Any]] {
            formats.append(contentsOf: adaptive)
        }

        for format in formats {
            guard let itag = format["itag"] as? NSNumber else { continue }
            if itag.stringValue == formatID {
                if let fps = format["fps"] as? NSNumber {
                    return Decimal(string: fps.stringValue)
                }
                return nil
            }
        }
        return nil
    }

    /// Parses `yt-dlp -j` output for the fallback path (spec §2.1):
    /// `formats[].format_id == formatID → fps`.
    public static func parseYtDlpInfo(_ data: Data, formatID: String) -> Decimal? {
        guard
            let object = try? JSONSerialization.jsonObject(with: data),
            let dict = object as? [String: Any],
            let formats = dict["formats"] as? [[String: Any]]
        else {
            return nil
        }
        for format in formats {
            let id: String?
            if let text = format["format_id"] as? String {
                id = text
            } else if let number = format["format_id"] as? NSNumber {
                id = number.stringValue
            } else {
                id = nil
            }
            guard id == formatID else { continue }
            if let fps = format["fps"] as? NSNumber {
                return Decimal(string: fps.stringValue)
            }
            return nil
        }
        return nil
    }
}
